zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/gateway/InclusiveGatewayTest.java

```diff
@@ -709,4 +714,541 @@ public void shouldEvaluateConditionWithoutDefaultFlow() {
+
+    // then
+    assertThat(
+            RecordingExporter.processInstanceRecords()
+                .withProcessInstanceKey(processInstanceKey)
+                .withIntent(ProcessInstanceIntent.ELEMENT_COMPLETED)
+                .limitToProcessInstanceCompleted())
+        .extracting(e -> e.getValue().getElementId())
+        .containsOnlyOnce("join");
+  }
+
+  @Test
+  public void shouldJoinGatewayIfActiveElementWithLoopNoPathToGateway() {
+    // given
+    final var process =
+        Bpmn.createExecutableProcess(PROCESS_ID)
+            .startEvent("start")
+            .parallelGateway("fork")
+            .task("task-1")
+            .inclusiveGateway("join")
+            .endEvent("end-1")
+            .moveToNode("fork")
+            .task("task-2")
+            .connectTo("join")
+            .moveToNode("fork")
+            .serviceTask("task-3")
+            .zeebeJobType("test")
+            .exclusiveGateway("exclusive")
+            .conditionExpression("a > 1")
+            .endEvent("end-2")
+            .moveToNode("exclusive")
+            .conditionExpression("a <= 1")
+            .connectTo("task-3")
+            .done();
+
+    ENGINE.deployment().withXmlResource(process).deploy();
+    final long processInstanceKey =
+        ENGINE.processInstance().ofBpmnProcessId(PROCESS_ID).withVariable("a", 5).create();
+
+    // when
+    ENGINE.job().ofInstance(processInstanceKey).withType("test").complete();
+
+    // then
+    assertThat(
+            RecordingExporter.processInstanceRecords()
+                .withProcessInstanceKey(processInstanceKey)
+                .limitToProcessInstanceCompleted()
+                .withIntent(ProcessInstanceIntent.ELEMENT_COMPLETED))
+        .extracting(e -> e.getValue().getElementId())
+        .containsOnlyOnce("join");
+  }
+
+  @Test
+  public void shouldJoinOnGatewayIfActiveLinkCatchEvent() {
+    // given
+    final ProcessBuilder processBuilder = Bpmn.createExecutableProcess(PROCESS_ID);
+    processBuilder
+        .startEvent()
+        .parallelGateway("fork")
+        .inclusiveGateway("join")
+        .endEvent()
+        .moveToNode("fork")
+        .intermediateThrowEvent("throw", b -> b.link("link"));
+
+    final BpmnModelInstance process =
+        processBuilder.linkCatchEvent("catch").link("link").connectTo("join").done();
+
+    ENGINE.deployment().withXmlResource(process).deploy();
+
+    // when
+    final long processInstanceKey = ENGINE.processInstance().ofBpmnProcessId(PROCESS_ID).create();
+
+    // then
+    assertThat(
+            RecordingExporter.processInstanceRecords()
+                .withProcessInstanceKey(processInstanceKey)
+                .limitToProcessInstanceCompleted()
+                .withElementId("join")
+                .withIntent(ProcessInstanceIntent.ELEMENT_COMPLETED))
+        .extracting(e -> e.getValue().getElementId(), Record::getIntent)
+        .containsOnlyOnce(tuple("join", ProcessInstanceIntent.ELEMENT_COMPLETED));
```

Comment:
:wrench: We should extend the verification to ensure that the process instance waits at the inclusive gateway. 

```suggestion
    assertThat(
            RecordingExporter.processInstanceRecords()
                .withProcessInstanceKey(processInstanceKey)
                .limitToProcessInstanceCompleted())
        .extracting(e -> e.getValue().getElementId(), Record::getIntent)
        .containsSubsequence(
            tuple("task-1", ProcessInstanceIntent.ELEMENT_COMPLETED),
            tuple("join", ProcessInstanceIntent.ELEMENT_ACTIVATING))
        .containsSubsequence(
            tuple("task-2", ProcessInstanceIntent.ELEMENT_COMPLETED),
            tuple("join", ProcessInstanceIntent.ELEMENT_ACTIVATING))
        .containsOnlyOnce(
            tuple("join", ProcessInstanceIntent.ELEMENT_COMPLETED),
            tuple(PROCESS_ID, ProcessInstanceIntent.ELEMENT_COMPLETED));
```

Reply:
:x: The test case fails with the updated process and verifications. 

```
Expecting actual:
  [("process", ACTIVATE_ELEMENT),
    ("process", ELEMENT_ACTIVATING),
    ("process", ELEMENT_ACTIVATED),
    ("startEvent_41f087b3-4a7a-49cc-820d-edc51b866932", ACTIVATE_ELEMENT),
    ("startEvent_41f087b3-4a7a-49cc-820d-edc51b866932", ELEMENT_ACTIVATING),
    ("startEvent_41f087b3-4a7a-49cc-820d-edc51b866932", ELEMENT_ACTIVATED),
    ("startEvent_41f087b3-4a7a-49cc-820d-edc51b866932", COMPLETE_ELEMENT),
    ("startEvent_41f087b3-4a7a-49cc-820d-edc51b866932", ELEMENT_COMPLETING),
    ("startEvent_41f087b3-4a7a-49cc-820d-edc51b866932", ELEMENT_COMPLETED),
    ("sequenceFlow_6d79d12e-9519-4063-8c46-c72d2f192522", SEQUENCE_FLOW_TAKEN),
    ("fork", ACTIVATE_ELEMENT),
    ("fork", ELEMENT_ACTIVATING),
    ("fork", ELEMENT_ACTIVATED),
    ("fork", ELEMENT_COMPLETING),
    ("fork", ELEMENT_COMPLETED),
    ("sequenceFlow_5ce126ed-a8d1-4ba4-a4c1-d6295507714b", SEQUENCE_FLOW_TAKEN),
    ("task-2", ACTIVATE_ELEMENT),
    ("sequenceFlow_f1a03bab-2312-461b-acaa-a5872194f8f9", SEQUENCE_FLOW_TAKEN),
    ("task-1", ACTIVATE_ELEMENT),
    ("task-2", ELEMENT_ACTIVATING),
    ("task-2", ELEMENT_ACTIVATED),
    ("task-1", ELEMENT_ACTIVATING),
    ("task-1", ELEMENT_ACTIVATED),
    ("task-1", COMPLETE_ELEMENT),
    ("task-1", ELEMENT_COMPLETING),
    ("task-1", ELEMENT_COMPLETED),
    ("sequenceFlow_3699d2cf-9108-4f14-b2f7-09b0b74a2952", SEQUENCE_FLOW_TAKEN),
    ("join", ACTIVATE_ELEMENT),
    ("join", ELEMENT_ACTIVATING),
    ("join", ELEMENT_ACTIVATED),
    ("join", ELEMENT_COMPLETING),
    ("join", ELEMENT_COMPLETED),
    ("sequenceFlow_d539266a-c78a-4998-b560-36bfc2514994", SEQUENCE_FLOW_TAKEN),
    ("endEvent_4537d4bb-cfe7-4d50-8ee2-fa119214d3c8", ACTIVATE_ELEMENT),
    ("endEvent_4537d4bb-cfe7-4d50-8ee2-fa119214d3c8", ELEMENT_ACTIVATING),
    ("endEvent_4537d4bb-cfe7-4d50-8ee2-fa119214d3c8", ELEMENT_ACTIVATED),
    ("endEvent_4537d4bb-cfe7-4d50-8ee2-fa119214d3c8", ELEMENT_COMPLETING),
    ("endEvent_4537d4bb-cfe7-4d50-8ee2-fa119214d3c8", ELEMENT_COMPLETED),
    ("task-2", COMPLETE_ELEMENT),
    ("task-2", ELEMENT_COMPLETING),
    ("task-2", ELEMENT_COMPLETED),
    ("sequenceFlow_9800ecc9-fc67-4000-989d-abf6d41ade15", SEQUENCE_FLOW_TAKEN),
    ("throw", ACTIVATE_ELEMENT),
    ("throw", ELEMENT_ACTIVATING),
    ("throw", ELEMENT_ACTIVATED),
    ("throw", COMPLETE_ELEMENT),
    ("throw", ELEMENT_COMPLETING),
    ("throw", ELEMENT_COMPLETED),
    ("catch", ACTIVATE_ELEMENT),
    ("catch", ELEMENT_ACTIVATING),
    ("catch", ELEMENT_ACTIVATED),
    ("catch", COMPLETE_ELEMENT),
    ("catch", ELEMENT_COMPLETING),
    ("catch", ELEMENT_COMPLETED),
    ("sequenceFlow_54f9d4b2-0109-40d8-992f-1090b2098f09", SEQUENCE_FLOW_TAKEN),
    ("join", ACTIVATE_ELEMENT),
    ("join", ELEMENT_ACTIVATING),
    ("join", ELEMENT_ACTIVATED),
    ("join", ELEMENT_COMPLETING),
    ("join", ELEMENT_COMPLETED),
    ("sequenceFlow_d539266a-c78a-4998-b560-36bfc2514994", SEQUENCE_FLOW_TAKEN),
    ("endEvent_4537d4bb-cfe7-4d50-8ee2-fa119214d3c8", ACTIVATE_ELEMENT),
    ("endEvent_4537d4bb-cfe7-4d50-8ee2-fa119214d3c8", ELEMENT_ACTIVATING),
    ("endEvent_4537d4bb-cfe7-4d50-8ee2-fa119214d3c8", ELEMENT_ACTIVATED),
    ("endEvent_4537d4bb-cfe7-4d50-8ee2-fa119214d3c8", ELEMENT_COMPLETING),
    ("endEvent_4537d4bb-cfe7-4d50-8ee2-fa119214d3c8", ELEMENT_COMPLETED),
    ("process", COMPLETE_ELEMENT),
    ("process", ELEMENT_COMPLETING),
    ("process", ELEMENT_COMPLETED)]
to contain only once:
  [("join", ELEMENT_COMPLETED), ("process", ELEMENT_COMPLETED)]
but some elements were found more than once:
  [("join", ELEMENT_COMPLETED)]
```

The inclusive gateway is completed two times. The process instance doesn't wait at the inclusive gateway until the second token arrives.